craft_parts/plugins/maven_plugin.py

```diff
@@ -74,6 +77,97 @@ def validate_environment(
+        effective_pom_path = Path("./effective.pom")
+        try:
+            self._execute(
+                # Write to file since writing to stdout without maven info logs require sudo, i.e.
+                # sudo mvn help:effective-pom -q -DforceStdout -doutput=/dev/stdout
+                f"{maven_executable} help:effective-pom -Doutput={effective_pom_path}"
+            )
+            project_java_major_version = _parse_project_java_version(effective_pom_path)
+            return _extract_java_version(project_java_major_version)
+        except subprocess.CalledProcessError as err:
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason=f"failed to generate effective pom: {err}",
+            ) from err
+        finally:
+            effective_pom_path.unlink(missing_ok=True)
+
+
+def _parse_project_java_version(effective_pom_path: Path) -> str | None:
+    """Parse the contents of effective pom XML file.
+
+    The order in which Maven determines the Java version is as follows:
+    1. maven-compiler-plugin's configuration release value
+    2. maven-compiler-plugin's property's release value
+    3. java.version property's value (Spring Boot specific)
+    """
+    tree = defusedxml.ElementTree.parse(effective_pom_path)
+    root = tree.getroot()
+    java_version_element: ET.Element | None = root.find(".//{*}java.version")
+    maven_compiler_release_element: ET.Element | None = root.find(
+        ".//{*}maven.compiler.release"
+    )
+    maven_compiler_plugin_release_element: ET.Element | None = None
+    plugins_element = root.find(".//{*}plugins")
+    if plugins_element is not None:
+        plugins = plugins_element.findall(".//{*}plugin")
+        release_element: ET.Element | None
+        for plugin in plugins:
+            if (
+                (artifact_id_element := plugin.find(".//{*}artifactId")) is not None
+                and artifact_id_element.text == "maven-compiler-plugin"
+                and (release_element := plugin.find(".//{*}release")) is not None
+                and release_element.text
+            ):
+                maven_compiler_plugin_release_element = release_element
+                break
+    if (
+        maven_compiler_plugin_release_element is not None
+        and maven_compiler_plugin_release_element.text
+    ):
+        return maven_compiler_plugin_release_element.text
+    if (
+        maven_compiler_release_element is not None
+        and maven_compiler_release_element.text
+    ):
+        return maven_compiler_release_element.text
+    if java_version_element is not None and java_version_element.text:
+        return java_version_element.text
+    return None
+
+
+def _extract_java_version(java_version_output: str | None) -> int | None:
+    if java_version_output is None:
+        return None
+    match = re.match(r"(1\.(\d+)|(\d+))", java_version_output)
```

Comment:
Similar to before, I think you can restructure this regex so that you always get it in the same group